fastestimator/trace/io/model_eval.py

```diff
@@ -0,0 +1,446 @@
+    """
+    def __init__(self, description: str, criteria: Callable, aggregate: bool = True, fail_threshold: int = 0) -> None:
+        self.description = description
+        self.criteria = criteria
+        self.criteria_inputs = inspect.signature(criteria).parameters.keys()
+        self.aggregate = aggregate
+        if not self.aggregate:
+            self.fail_threshold = fail_threshold
+
+    def clean_result(self) -> None:
+        if not self.aggregate:
+            self.result = []
+            self.fail_id = []
+
+
+@traceable()
+class ModelEval(Trace):
+    """Automate testing and report generation.
+
+    Args:
+        test_cases: List of TestCase object.
+        save_path: Where to save the output directory
+        test_title: Title of the test.
+        data_id: Data instance ID key. If provided, then sample-case test will return failing sample ID.
+    """
+    def __init__(self,
+                 test_cases: Union[TestCase, List[TestCase]],
+                 save_path: str,
+                 test_title: str = "Test",
+                 data_id: str = None) -> None:
+
+        self.test_title = test_title
+        self.test_cases = to_list(test_cases)
+        self.instance_cases = []
+        self.aggregate_cases = []
+        self.sample_id = data_id
+
+        all_inputs = set()
+        for case in self.test_cases:
+            all_inputs.update(case.criteria_inputs)
+            case.clean_result()
+            if case.aggregate:
+                self.aggregate_cases.append(case)
+            else:
+                self.instance_cases.append(case)
+
+        if self.sample_id:
+            all_inputs.update([data_id])
+
+        path = os.path.normpath(save_path)
+        path = os.path.abspath(path)
+        root_dir = os.path.dirname(path)
+        report = os.path.basename(path) or 'report'
+        report = report.split('.')[0]
+        self.save_dir = os.path.join(root_dir, report)
+        self.src_dir = os.path.join(self.save_dir, "resources")
+        os.makedirs(self.save_dir, exist_ok=True)
+        os.makedirs(self.src_dir, exist_ok=True)
+
+        super().__init__(inputs=all_inputs, mode="test")
+
+    def on_begin(self, data: Data) -> None:
+        self._sanitize_report_name()
+        self._initialize_json_summary()
+
+    def on_batch_end(self, data: Data) -> None:
+        for case in self.instance_cases:
+            result = case.criteria(*[data[var_name] for var_name in case.criteria_inputs])
+            if not isinstance(result, np.ndarray):
+                raise TypeError("Criteria return of per-instance test need to be ndarray with dtype bool")
+            elif result.dtype != np.dtype("bool"):
+                raise TypeError("Criteria return of per-instance test need to be ndarray with dtype bool")
+
+            result = result.reshape(-1)
+            case.result.append(result)
+            if self.sample_id:
+                data_id = to_number(data[self.sample_id]).reshape((-1, ))
```

Comment:
why are we assuming that data_id must be a number?

Reply:
It doesn't assume data_id to be a number. It just converts it from tensor to Numpy array. The Numpy array can still contain strings

Reply:
I indeed hadn't tested with string ID, and later did find a couple of errors in that scenario.
Now it can work with string ID.